File: scripts/sync_inventory.py
#!/usr/bin/env python3
"""sync_inventory.py — Python equivalent of scripts/sync-inventory.mjs.

Exists because some run environments (e.g. the unattended Cowork routine) have no
Node. Verified to produce output semantically identical to the .mjs (same SKU set,
same lot data; 0 diffs) from the same CSV. Prefer the .mjs when Node is available;
use this otherwise. Same flags: [--in <csv>] [--out <name>] [--check] [--promote].

Touches ONLY inventory. Never reads or writes pricing (catalog.json).
"""
import sys, os, re, json, math, glob, time, shutil
from datetime import datetime, timezone

SCRIPTS_DIR = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.dirname(SCRIPTS_DIR)
DATA_DIR = os.path.join(REPO, "src/data/montitrentini")
SRC_DIR = os.path.join(DATA_DIR, "source")
args = sys.argv[1:]

def get_arg(flag):
    if flag in args:
        i = args.index(flag)
        return args[i + 1] if i + 1 < len(args) else None
    return None

def newest_csv():
    files = sorted(os.path.basename(f) for f in glob.glob(os.path.join(SRC_DIR, "*"))
                   if re.match(r'^availability_.*\.csv$', os.path.basename(f), re.I))
    return os.path.join(SRC_DIR, files[-1]) if files else None

IN = get_arg("--in") or newest_csv()
out_flag = get_arg("--out")
OUT = os.path.join(DATA_DIR, out_flag) if out_flag else os.path.join(DATA_DIR, "inventory.json")
if not IN or not os.path.exists(IN):
    sys.stderr.write("No input CSV found. Pass --in <path> or place source/availability_*.csv\n"); sys.exit(1)

def parse_csv(text):
    rows, row, field, inQ = [], [], "", False
    i, n = 0, len(text)
    while i < n:
        c = text[i]
        if inQ:
            if c == '"':
                if i + 1 < n and text[i + 1] == '"':
                    field += '"'; i += 1
                else:
                    inQ = False
            else:
                field += c
        elif c == '"':
            inQ = True
        elif c == ",":
            row.append(field); field = ""
        elif c == "\n":
            row.append(field); rows.append(row); row = []; field = ""
        elif c == "\r":
            pass
        else:
            field += c
        i += 1
    if field or row:
        row.append(field); rows.append(row)
    return rows

MONTHS = {"jan":"01","feb":"02","mar":"03","apr":"04","may":"05","jun":"06",
          "jul":"07","aug":"08","sep":"09","oct":"10","nov":"11","dec":"12"}
TODAY = datetime.now()

def clean(s):
    return "" if s is None else str(s).strip()

def iso_from_mdy(s):
    m = re.match(r'^(\d{1,2})/(\d{1,2})/(\d{4})$', clean(s))
    if not m: return None
    return f"{m.group(3)}-{m.group(1).zfill(2)}-{m.group(2).zfill(2)}"

def iso_from_eta(s):
    m = re.search(r'(\d{1,2})/(\d{1,2})(?:/(\d{2,4}))?', clean(s))
    if not m: return None
    mo, d, g3 = int(m.group(1)), int(m.group(2)), m.group(3)
    if g3:
        y = 2000 + int(g3) if len(g3) == 2 else int(g3)
    else:
        y = TODAY.year
        if (TODAY - datetime(y, mo, d)).total_seconds() / 86400 > 183:
            y += 1
    return f"{y}-{str(mo).zfill(2)}-{str(d).zfill(2)}"

def month_year(s):
    m = re.match(r'^([A-Za-z]{3})-(\d{2})$', clean(s))
    if not m: return None
    mo = MONTHS.get(m.group(1).lower())
    return f"20{m.group(2)}-{mo}" if mo else None

def num(s):
    t = clean(s).replace(",", "")
    if t == "": return None
    try:
        n = float(t)
    except ValueError:
        return None
    if not math.isfinite(n): return None
    return int(n) if n == int(n) else n

def int_or_0(s):
    n = num(s)
    return 0 if n is None else math.trunc(n)

raw = open(IN, "r", encoding="utf-8").read()
rows = parse_csv(raw)

last_updated = TODAY.strftime("%Y-%m-%d")
banner = next((r for r in rows if any(re.search(r'updated on', c, re.I) for c in r)), None)
if banner:
    idx = next(i for i, c in enumerate(banner) if re.search(r'updated on', c, re.I))
    val = clean(banner[idx + 1]) if idx + 1 < len(banner) else ""
    m = re.search(r'(\d{1,2})\s+([A-Za-z]+)\s+(\d{4})', val)
    if m:
        mo = MONTHS.get(m.group(2)[:3].lower())
        if mo:
            last_updated = f"{m.group(3)}-{mo}-{m.group(1).zfill(2)}"

h_idx = next((i for i, r in enumerate(rows) if clean(r[0]).lower() == "item"), -1)
data_rows = rows[h_idx + 1:]

skus = {}
def ensure(code, name):
    code = clean(code)
    if not code: return None
    if code not in skus:
        skus[code] = {"code": code, "name": clean(name), "casesAvail": 0,
                      "casesInTransit": 0, "comment": "", "commentEn": "", "lots": []}
    elif not skus[code]["name"] and name:
        skus[code]["name"] = clean(name)
    return skus[code]

def cell(r, i):
    return r[i] if i < len(r) else ""

for r in data_rows:
    code = clean(cell(r, 0))
    if not code: continue
    s = ensure(code, cell(r, 1))
    s["casesAvail"] = int_or_0(cell(r, 2))
    s["comment"] = clean(cell(r, 6))

lot_count = 0
for r in data_rows:
    code = clean(cell(r, 8))
    if not code: continue
    s = ensure(code, cell(r, 9))
    receipt_raw = clean(cell(r, 11))
    receipt_date = iso_from_mdy(receipt_raw)
    on_hand = bool(receipt_date)
    eta = exp_date = exp_month = None
    if on_hand:
        exp_date = iso_from_mdy(clean(cell(r, 16)))
    else:
        eta = iso_from_eta(receipt_raw)
        exp_month = month_year(clean(cell(r, 16)))
    net_avail = num(cell(r, 15))
    s["lots"].append({
        "lotNum": clean(cell(r, 10)) or None,
        "status": "on_hand" if on_hand else "in_transit",
        "receiptDate": receipt_date or None,
        "eta": eta,
        "cases": int_or_0(cell(r, 12)),
        "reserved": int_or_0(cell(r, 13)),
        "netAvailLb": None if net_avail is None else net_avail,
        "expDate": exp_date,
        "expMonth": exp_month or None,
        "comment": clean(cell(r, 14)),
    })
    lot_count += 1

for code in list(skus.keys()):
    s = skus[code]
    s["casesInTransit"] = sum((l["cases"] or 0) for l in s["lots"] if l["status"] == "in_transit")

out = {
    "schemaVersion": "1.2",
    "clientId": "monti-trentini",
    "lastUpdated": last_updated,
    "source": os.path.basename(IN),
    "generatedAt": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.")
                   + f"{datetime.now(timezone.utc).microsecond // 1000:03d}Z",
    "generatedBy": "scripts/sync-inventory.mjs",
    "skus": skus,
}

sku_count = len(skus)
with_stock = sum(1 for s in skus.values()
                 if any(l["status"] == "on_hand" and (l["cases"] - l["reserved"]) > 0 for l in s["lots"]))

def validate(doc):
    errs = []
    n = len(doc["skus"])
    if n < 80: errs.append(f"only {n} SKUs parsed (expected ~110+) — sheet shape may have changed")
    if lot_count < 80: errs.append(f"only {lot_count} lots parsed (expected ~120) — lot table may be malformed")
    if with_stock < 10: errs.append(f"only {with_stock} sellable-now SKUs (expected ~40) — suspiciously low")
    for code, s in doc["skus"].items():
        for l in s["lots"]:
            if l["status"] == "on_hand" and not l["expDate"]:
                errs.append(f"on-hand lot {code}/{l['lotNum']} missing expDate")
            if l["cases"] is None:
                errs.append(f"lot {code}/{l['lotNum']} has no case count")
    return errs[:8]

CANON = os.path.join(DATA_DIR, "inventory.json")
def sellable_map(doc):
    return {k: sum(max(0, (l["cases"] or 0) - (l["reserved"] or 0))
                   for l in s["lots"] if l["status"] == "on_hand")
            for k, s in doc["skus"].items()}

def diff_vs_canon(doc):
    if not os.path.exists(CANON):
        return {"changed": True, "lines": ["(no existing inventory.json — first promote)"]}
    try:
        prev = json.load(open(CANON, encoding="utf-8"))
    except Exception:
        return {"changed": True, "lines": ["(existing inventory.json unreadable)"]}
    a, b = sellable_map(prev), sellable_map(doc)
    keys = list(dict.fromkeys(list(a.keys()) + list(b.keys())))
    lines = []
    for k in keys:
        x, y = a.get(k), b.get(k)
        if x != y:
            nm = (doc["skus"].get(k) or prev["skus"].get(k))["name"]
            lines.append(f"{k} {nm}: {x if x is not None else '-'} -> {y if y is not None else '-'}")
    return {"changed": len(lines) > 0 or prev.get("lastUpdated") != doc["lastUpdated"], "lines": lines}

def dump(doc):
    return json.dumps(doc, ensure_ascii=False, indent=2) + "\n"

if "--promote" in args or "--check" in args:
    errs = validate(out)
    diff = diff_vs_canon(out)
    if errs:
        sys.stderr.write("✗ VALIDATION FAILED — not promoting:\n")
        for e in errs: sys.stderr.write("   - " + e + "\n")
        sys.exit(2)
    if not diff["changed"]:
        print(f"= No change vs current inventory.json (sheet lastUpdated {last_updated}). Nothing to promote.")
        sys.exit(3)
    if "--check" in args:
        print(f"✓ VALID & CHANGED ({len(diff['lines'])} SKU sellable-now deltas). Safe to promote.")
        for l in diff["lines"][:12]: print("   " + l)
        sys.exit(0)
    if os.path.exists(CANON):
        bdir = os.path.abspath(os.path.join(DATA_DIR, f"../../archive/backup_{TODAY.strftime('%Y-%m-%d')}_inventory_autosync"))
        os.makedirs(bdir, exist_ok=True)
        shutil.copyfile(CANON, os.path.join(bdir, f"inventory_{int(time.time() * 1000)}.json"))
    open(CANON, "w", encoding="utf-8").write(dump(out))
    print(f"✓ PROMOTED -> {os.path.relpath(CANON)}  (SKUs {sku_count} | lots {lot_count} | sellable-now {with_stock})")
    for l in diff["lines"][:12]: print("   " + l)
    sys.exit(0)

open(OUT, "w", encoding="utf-8").write(dump(out))
print(f"✓ {os.path.basename(IN)} -> {os.path.relpath(OUT)}")
print(f"  SKUs: {sku_count} | lots: {lot_count} | sellable-now SKUs: {with_stock} | lastUpdated: {last_updated}")

File: scripts/sync-inventory.mjs
#!/usr/bin/env node
/* sync-inventory.mjs — regenerate inventory.json from the Monti Trentini
   "Availability of items and pending orders" sheet (exported to CSV).

   Pipeline (Path A, session-driven): Claude exports the live Google Sheet to
   src/data/montitrentini/source/availability_<YYYY-MM-DD>.csv, then runs this
   script to produce inventory.json in the canonical schema (v1.2) that
   pricing-core.js consumes (skus[code].lots[] with status/expDate/reserved).

   The sheet's first tab holds TWO side-by-side tables:
     cols 0..6  = per-SKU summary  (Item, Description, Cases Available,
                  Suggested USA, Suggested ITA, Confirmed, Comments)
     col  7     = spacer
     cols 8..16 = lot detail (Item, Description, Lot#, Receipt Date,
                  Cases Available, Reserved, Comments, Net Available, Expiration)

   Usage:
     node scripts/sync-inventory.mjs [--in <csv>] [--out <json>] [--quiet]
                                     [--check | --promote] [--require-drive-meta]
   --require-drive-meta: hard-fail (exit 4) if the Drive sidecar is missing, so an
   unattended run can never publish the sheet's hand-typed date. Use it in cron.
   Defaults: newest source/availability_*.csv  ->  inventory.json
   Safe-by-default for review: pass --out inventory.NEW.json to avoid replacing.
*/
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DATA_DIR = path.resolve(__dirname, "../src/data/montitrentini");
const SRC_DIR = path.join(DATA_DIR, "source");

// ---- args -----------------------------------------------------------------
const args = process.argv.slice(2);
const getArg = (flag) => { const i = args.indexOf(flag); return i >= 0 ? args[i + 1] : null; };
const QUIET = args.includes("--quiet");
const log = (...a) => { if (!QUIET) console.log(...a); };

function newestCsv() {
  if (!fs.existsSync(SRC_DIR)) return null;
  const files = fs.readdirSync(SRC_DIR).filter((f) => /^availability_.*\.csv$/i.test(f)).sort();
  return files.length ? path.join(SRC_DIR, files[files.length - 1]) : null;
}
const IN = getArg("--in") || newestCsv();
const OUT = path.resolve(getArg("--out") ? path.join(DATA_DIR, getArg("--out")) : path.join(DATA_DIR, "inventory.json"));
if (!IN || !fs.existsSync(IN)) { console.error("No input CSV found. Pass --in <path> or place source/availability_*.csv"); process.exit(1); }

// ---- tiny CSV parser (handles quoted fields w/ commas + escaped quotes) ----
function parseCsv(text) {
  const rows = [];
  let row = [], field = "", inQ = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inQ) {
      if (c === '"') { if (text[i + 1] === '"') { field += '"'; i++; } else inQ = false; }
      else field += c;
    } else if (c === '"') inQ = true;
    else if (c === ",") { row.push(field); field = ""; }
    else if (c === "\n") { row.push(field); rows.push(row); row = []; field = ""; }
    else if (c === "\r") { /* skip */ }
    else field += c;
  }
  if (field.length || row.length) { row.push(field); rows.push(row); }
  return rows;
}

// ---- helpers --------------------------------------------------------------
const MONTHS = { jan:"01", feb:"02", mar:"03", apr:"04", may:"05", jun:"06",
  jul:"07", aug:"08", sep:"09", oct:"10", nov:"11", dec:"12" };
const TODAY = new Date();

const clean = (s) => (s == null ? "" : String(s).trim());

// "03/23/2026" -> "2026-03-23"
function isoFromMDY(s) {
  const m = clean(s).match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/);
  if (!m) return null;
  const [, mo, d, y] = m;
  return `${y}-${mo.padStart(2, "0")}-${d.padStart(2, "0")}`;
}
// "ETA 07/02" (no year) -> "2026-07-02", inferring nearest sensible year
function isoFromEta(s) {
  const m = clean(s).match(/(\d{1,2})\/(\d{1,2})(?:\/(\d{2,4}))?/);
  if (!m) return null;
  const mo = +m[1], d = +m[2];
  let y = m[3] ? (m[3].length === 2 ? 2000 + +m[3] : +m[3]) : TODAY.getFullYear();
  if (!m[3]) {
    // ETA is a near-future date; if it lands >6 months before today, roll forward
    const cand = new Date(y, mo - 1, d);
    if ((TODAY - cand) / 86400000 > 183) y += 1;
  }
  return `${y}-${String(mo).padStart(2, "0")}-${String(d).padStart(2, "0")}`;
}
// "May-27" / "Feb-27" -> "2027-05" (month-year only; kept for display)
function monthYear(s) {
  const m = clean(s).match(/^([A-Za-z]{3})-(\d{2})$/);
  if (!m) return null;
  const mo = MONTHS[m[1].toLowerCase()];
  return mo ? `20${m[2]}-${mo}` : null;
}
const num = (s) => {
  const t = clean(s).replace(/,/g, "");
  if (t === "") return null;
  const n = Number(t);
  return Number.isFinite(n) ? n : null;
};
const intOr0 = (s) => { const n = num(s); return n == null ? 0 : Math.trunc(n); };

// ---- read & split ---------------------------------------------------------
const raw = fs.readFileSync(IN, "utf8");
const rows = parseCsv(raw);

// capture "Updated on: 18 June 2026 16:47" from the banner row.
// The banner shows up in two formats depending on how MT typed it that week:
//   written-out, day-first  : "28 July 2026"      (European style)
//   numeric, month-first    : "7/30/2026 12:00 PM" (American style, M/D/YYYY)
// Both must resolve to the SAME ISO date convention (YYYY-MM-DD) before this
// value is ever compared against anything else. Get the numeric case wrong —
// e.g. read "7/30" as day=7/month=30 the European way — and it either throws
// (no month 30) or silently produces a bogus date that then falsely disagrees
// with Drive's real mtime. So: numeric banner dates are ALWAYS interpreted
// American (first number = month), matching how Monti Trentini's team writes
// them. This translation must happen right here, before lastUpdated is used
// for any downstream comparison ("date authentication").
let lastUpdated = TODAY.toISOString().slice(0, 10);
let bannerParsed = false;
const banner = rows.find((r) => r.some((c) => /updated on/i.test(c)));
if (banner) {
  const idx = banner.findIndex((c) => /updated on/i.test(c));
  const val = clean(banner[idx + 1]);

  // 1) written-out day-first form: "28 July 2026"
  const written = val.match(/(\d{1,2})\s+([A-Za-z]+)\s+(\d{4})/);
  if (written) {
    const mo = MONTHS[written[2].slice(0, 3).toLowerCase()];
    if (mo) {
      lastUpdated = `${written[3]}-${mo}-${written[1].padStart(2, "0")}`;
      bannerParsed = true;
    }
  }

  // 2) numeric American form: "7/30/2026" or "7/30/2026 12:00 PM"
  //    First number is the MONTH, not the day — do not swap these.
  if (!bannerParsed) {
    const numeric = val.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})/);
    if (numeric) {
      const [, moStr, dStr, yStr] = numeric;
      const mo = Number(moStr), d = Number(dStr);
      if (mo >= 1 && mo <= 12 && d >= 1 && d <= 31) {
        lastUpdated = `${yStr}-${String(mo).padStart(2, "0")}-${String(d).padStart(2, "0")}`;
        bannerParsed = true;
      }
    }
  }

  if (!bannerParsed) {
    console.warn(`! Banner text "${val}" didn't match a known date format — lastUpdated defaulted to today (${lastUpdated}). Add a case for this format instead of trusting the default.`);
  }
}

// ---- authoritative "last updated": Google Drive modifiedTime --------------
// The banner cell above is hand-typed on the weekly refresh, which fails twice:
// it misses the intermittent mid-week corrections MT makes, and it carries
// typos straight into the buyer catalog. On 2026-07-25 it read "28 July 2026"
// — three days in the future — while Drive's modifiedTime for the sheet was
// 2026-07-24T15:53:09Z. A future date also makes agency-console's stock-age
// calculation go negative.
//
// Drive's timestamp is the source of truth. It reaches this script through a
// sidecar written at export time, so the script stays runnable offline with no
// Drive credentials:
//     source/availability_<YYYY-MM-DD>.meta.json
//     { "driveFileId": "1meZQQ…", "driveModifiedTime": "2026-07-24T15:53:09.134Z",
//       "sheetOwner": "order@montitrentini-usa.com", "exportedAt": "…" }
// Absent the sidecar it falls back to the banner and says so.
//
// NOTE: this Drive connector exposes modifiedTime and the file OWNER, but not
// lastModifyingUser — so "when" is exact, "who" is the sheet owner, not the
// individual editor. Wiring the Sheets revisions API would close that gap.
const localDate = (iso) =>
  new Date(iso).toLocaleDateString("en-CA", { timeZone: "America/New_York" });

const sheetStatedUpdate = lastUpdated;          // whatever the banner claimed
let lastUpdatedSource = "sheet-banner";
let driveModifiedTime = null, driveFileId = null, sheetOwner = null;

const META = IN.replace(/\.csv$/i, ".meta.json");
if (fs.existsSync(META)) {
  try {
    const meta = JSON.parse(fs.readFileSync(META, "utf8"));
    if (meta.driveModifiedTime) {
      driveModifiedTime = meta.driveModifiedTime;
      driveFileId = meta.driveFileId || null;
      sheetOwner = meta.sheetOwner || null;
      lastUpdated = localDate(driveModifiedTime);
      lastUpdatedSource = "drive-modifiedTime";
    }
  } catch (e) {
    console.warn(`! ${path.basename(META)} unreadable (${e.message}) — falling back to the sheet banner.`);
  }
} else {
  console.warn(`! No ${path.basename(META)} beside the CSV — using the hand-typed banner (${sheetStatedUpdate}).`);
  console.warn(`  Write the sidecar at export time so mid-week corrections aren't missed.`);
}

// Unattended runs must never fall back to the banner in silence. The scheduled
// "Monti inventory watch" runs at 08:00 with nobody reading stdout, and a fresh
// session may not have the Google Drive connector available at all — in which
// case no sidecar can be written. Pass --require-drive-meta there so the run
// STOPS rather than publishing a hand-typed (possibly typo'd) date to the live
// buyer catalog. Exit 4 = "Drive metadata required but unavailable".
if (args.includes("--require-drive-meta") && lastUpdatedSource !== "drive-modifiedTime") {
  console.error("");
  console.error("x DRIVE METADATA REQUIRED BUT MISSING — refusing to continue.");
  console.error(`x   expected sidecar: ${path.relative(process.cwd(), META)}`);
  console.error("x   Without it lastUpdated falls back to the sheet's hand-typed banner,");
  console.error("x   which is exactly the value this flag exists to keep off the live catalog.");
  console.error("x   Write the sidecar at export time:");
  console.error('x     { "driveFileId": "...", "driveModifiedTime": "<ISO>", "sheetOwner": "..." }');
  console.error("");
  process.exit(4);
}

if (driveModifiedTime && sheetStatedUpdate !== lastUpdated) {
  const future = sheetStatedUpdate > TODAY.toISOString().slice(0, 10);
  const bar = "!".repeat(66);
  console.warn("");
  console.warn(bar);
  console.warn("!  SHEET BANNER DISAGREES WITH GOOGLE DRIVE");
  console.warn(`!    banner says : ${sheetStatedUpdate}${future ? "   <-- IN THE FUTURE, almost certainly a typo" : ""}`);
  console.warn(`!    Drive says  : ${lastUpdated}   (${driveModifiedTime})`);
  console.warn("!  Using Drive — the banner is typed by hand, Drive is the file's real mtime.");
  console.warn("!  A large gap means the sheet was edited without the banner being updated.");
  console.warn(bar);
  console.warn("");
}

// find header row (the one starting with "Item")
const hIdx = rows.findIndex((r) => clean(r[0]).toLowerCase() === "item");
const dataRows = rows.slice(hIdx + 1);

const skus = {};
function ensure(code, name) {
  code = clean(code);
  if (!code) return null;
  if (!skus[code]) skus[code] = { code, name: clean(name), casesAvail: 0, casesInTransit: 0, comment: "", commentEn: "", lots: [] };
  else if (!skus[code].name && name) skus[code].name = clean(name);
  return skus[code];
}

// Pass 1: summary table (cols 0..6) — establishes the SKU roster + headline + comment
for (const r of dataRows) {
  const code = clean(r[0]);
  if (!code) continue;
  const s = ensure(code, r[1]);
  s.casesAvail = intOr0(r[2]);
  s.comment = clean(r[6]);
}

// Pass 2: lot table (cols 8..16)
let lotCount = 0;
for (const r of dataRows) {
  const code = clean(r[8]);
  if (!code) continue;
  const s = ensure(code, r[9]);
  const lotNum = clean(r[10]) || null;
  const receiptRaw = clean(r[11]);
  const cases = intOr0(r[12]);
  const reserved = intOr0(r[13]);
  const comment = clean(r[14]);
  const netAvailLb = num(r[15]);
  const expRaw = clean(r[16]);

  const receiptDate = isoFromMDY(receiptRaw);
  const onHand = !!receiptDate; // a real MM/DD/YYYY receipt date == landed stock
  let eta = null, expDate = null, expMonth = null;
  if (onHand) {
    expDate = isoFromMDY(expRaw); // full date when on hand
  } else {
    eta = isoFromEta(receiptRaw); // "ETA 07/02" / "AIR FREIGHT" rows carry ETA here
    expMonth = monthYear(expRaw); // in-transit expiry is month-year only
  }

  const lot = {
    lotNum,
    status: onHand ? "on_hand" : "in_transit",
    receiptDate: receiptDate || null,
    eta,
    cases,
    reserved,
    netAvailLb: netAvailLb == null ? null : netAvailLb,
    expDate,                 // null for in-transit (matches schema v1.2 / allocate())
    expMonth: expMonth || null, // extra: month-year hint for in-transit (display only)
    comment,
  };
  s.lots.push(lot);
  lotCount++;
}

// Pass 3: derive casesInTransit per SKU from in-transit lots
for (const code of Object.keys(skus)) {
  const s = skus[code];
  s.casesInTransit = s.lots.filter((l) => l.status === "in_transit").reduce((a, l) => a + (l.cases || 0), 0);
}

const out = {
  schemaVersion: "1.2",
  clientId: "monti-trentini",
  lastUpdated,
  lastUpdatedSource,
  sheetStatedUpdate,
  sheetModifiedAt: driveModifiedTime,
  sheetFileId: driveFileId,
  sheetOwner,
  source: path.basename(IN),
  generatedAt: TODAY.toISOString(),
  generatedBy: "scripts/sync-inventory.mjs",
  skus,
};

const skuCount = Object.keys(skus).length;
const withStock = Object.values(skus).filter((s) => s.lots.some((l) => l.status === "on_hand" && (l.cases - l.reserved) > 0)).length;

// ---- validation gate (for unattended --promote runs) ----------------------
// Returns [] when safe to ship, or a list of human-readable failures.
// Cross-checks parsed items against source/item-reference.json (static code ->
// description truth list). A materially different description on a KNOWN code
// means the sheet's columns shifted or the parse mis-aligned:
//   >=5 description mismatches  -> hard fail (structural shift)
//   1-4 mismatches              -> warn, still promote (legit renames happen)
//   unknown codes               -> warn only (new items; add them to the reference)
const REF_PATH = path.join(SRC_DIR, "item-reference.json");
const normDesc = (s) => String(s || "").toUpperCase().replace(/[^A-Z0-9]+/g, " ").trim();
// "materially different": no significant word overlap after normalization
function descMatches(a, b) {
  const wa = new Set(normDesc(a).split(" ")), wb = new Set(normDesc(b).split(" "));
  let hit = 0;
  for (const w of wa) if (wb.has(w)) hit++;
  return hit >= Math.min(2, wa.size, wb.size); // share >=2 words (or all, if shorter)
}
function validate(doc) {
  const errs = [];
  const n = Object.keys(doc.skus).length;
  if (n < 80) errs.push(`only ${n} SKUs parsed (expected ~110+) — sheet shape may have changed`);
  if (lotCount < 80) errs.push(`only ${lotCount} lots parsed (expected ~120) — lot table may be malformed`);
  if (withStock < 10) errs.push(`only ${withStock} sellable-now SKUs (expected ~40) — suspiciously low`);
  for (const [code, s] of Object.entries(doc.skus)) {
    for (const l of s.lots) {
      if (l.status === "on_hand" && !l.expDate) errs.push(`on-hand lot ${code}/${l.lotNum} missing expDate`);
      if (l.cases == null) errs.push(`lot ${code}/${l.lotNum} has no case count`);
    }
  }
  // reference cross-check
  if (fs.existsSync(REF_PATH)) {
    let ref = null;
    try { ref = JSON.parse(fs.readFileSync(REF_PATH, "utf8")).items; } catch { /* unreadable -> skip */ }
    if (ref) {
      const mismatches = [], unknown = [];
      for (const [code, s] of Object.entries(doc.skus)) {
        if (!(code in ref)) { unknown.push(code); continue; }
        if (!descMatches(s.name, ref[code])) mismatches.push(`${code}: sheet "${s.name}" vs reference "${ref[code]}"`);
      }
      const missing = Object.keys(ref).filter((c) => !(c in doc.skus));
      if (mismatches.length >= 5) {
        errs.push(`${mismatches.length} item descriptions do not match item-reference.json — columns likely shifted`);
        for (const m of mismatches.slice(0, 4)) errs.push(`  ${m}`);
      } else if (mismatches.length) {
        for (const m of mismatches) log(`⚠ description drift (promoting anyway): ${m} — update item-reference.json if legit rename`);
      }
      if (missing.length >= 20) errs.push(`${missing.length} known codes vanished from the sheet (e.g. ${missing.slice(0, 5).join(", ")}) — parse likely broken`);
      if (unknown.length) log(`⚠ ${unknown.length} new code(s) not in item-reference.json: ${unknown.join(", ")} — add if legit new items`);
    }
  } else {
    log("⚠ no source/item-reference.json — description cross-check skipped");
  }
  return errs.slice(0, 8);
}

const CANON = path.join(DATA_DIR, "inventory.json");
function sellableMap(doc) {
  const m = {};
  for (const [k, s] of Object.entries(doc.skus))
    m[k] = s.lots.filter((l) => l.status === "on_hand").reduce((a, l) => a + Math.max(0, (l.cases || 0) - (l.reserved || 0)), 0);
  return m;
}
function diffVsCanon(doc) {
  if (!fs.existsSync(CANON)) return { changed: true, lines: ["(no existing inventory.json — first promote)"] };
  let prev; try { prev = JSON.parse(fs.readFileSync(CANON, "utf8")); } catch { return { changed: true, lines: ["(existing inventory.json unreadable)"] }; }
  const a = sellableMap(prev), b = sellableMap(doc);
  const keys = [...new Set([...Object.keys(a), ...Object.keys(b)])];
  const lines = [];
  for (const k of keys) {
    const x = a[k], y = b[k];
    if (x !== y) lines.push(`${k} ${(doc.skus[k] || prev.skus[k]).name}: ${x ?? "-"} -> ${y ?? "-"}`);
  }
  return { changed: lines.length > 0 || prev.lastUpdated !== doc.lastUpdated, lines };
}

if (args.includes("--promote") || args.includes("--check")) {
  const errs = validate(out);
  const diff = diffVsCanon(out);
  if (errs.length) {
    console.error("✗ VALIDATION FAILED — not promoting:");
    errs.forEach((e) => console.error("   - " + e));
    process.exit(2);
  }
  if (!diff.changed) {
    log(`= No change vs current inventory.json (sheet lastUpdated ${lastUpdated}). Nothing to promote.`);
    process.exit(3); // distinct code: "valid but unchanged"
  }
  if (args.includes("--check")) {
    log(`✓ VALID & CHANGED (${diff.lines.length} SKU sellable-now deltas). Safe to promote.`);
    diff.lines.slice(0, 12).forEach((l) => log("   " + l));
    process.exit(0);
  }
  // --promote: back up current, then replace
  if (fs.existsSync(CANON)) {
    const stamp = TODAY.toISOString().slice(0, 10);
    const bdir = path.resolve(DATA_DIR, `../../archive/backup_${stamp}_inventory_autosync`);
    fs.mkdirSync(bdir, { recursive: true });
    fs.copyFileSync(CANON, path.join(bdir, `inventory_${Date.now()}.json`));
  }
  fs.writeFileSync(CANON, JSON.stringify(out, null, 2) + "\n");
  log(`✓ PROMOTED -> ${path.relative(process.cwd(), CANON)}  (SKUs ${skuCount} | lots ${lotCount} | sellable-now ${withStock})`);
  diff.lines.slice(0, 12).forEach((l) => log("   " + l));
  process.exit(0);
}

fs.writeFileSync(OUT, JSON.stringify(out, null, 2) + "\n");
log(`✓ ${path.basename(IN)} -> ${path.relative(process.cwd(), OUT)}`);
log(`  SKUs: ${skuCount} | lots: ${lotCount} | sellable-now SKUs: ${withStock}`);
log(`  lastUpdated: ${lastUpdated}  (source: ${lastUpdatedSource}${driveModifiedTime ? `, sheet banner said ${sheetStatedUpdate}` : ""})`);
